frame/support/procedural/src/derive_impl.rs

```diff
@@ -0,0 +1,183 @@
+		&self.foreign_path
+	}
+}
+
+impl ToTokens for DeriveImplAttrArgs {
+	fn to_tokens(&self, tokens: &mut TokenStream2) {
+		tokens.extend(self.foreign_path.to_token_stream());
+		tokens.extend(self._as.to_token_stream());
+		tokens.extend(self.disambiguation_path.to_token_stream());
+	}
+}
+
+/// Gets the [`Ident`] representation of the given [`ImplItem`], if one exists. Otherwise
+/// returns [`None`].
+///
+/// Used by [`combine_impls`] to determine whether we can compare [`ImplItem`]s by [`Ident`]
+/// or not.
+fn impl_item_ident(impl_item: &ImplItem) -> Option<&Ident> {
+	match impl_item {
+		ImplItem::Const(item) => Some(&item.ident),
+		ImplItem::Fn(item) => Some(&item.sig.ident),
+		ImplItem::Type(item) => Some(&item.ident),
+		ImplItem::Macro(item) => item.mac.path.get_ident(),
+		_ => None,
+	}
+}
+
+/// The real meat behind `derive_impl`. Takes in a `local_impl`, which is the impl for which we
+/// want to implement defaults (i.e. the one the attribute macro is attached to), and a
+/// `foreign_impl`, which is the impl containing the defaults we want to use, and returns an
+/// [`ItemImpl`] containing the final generated impl.
+///
+/// This process has the following caveats:
+/// * Colliding items that have an ident are not copied into `local_impl`
+/// * Uncolliding items that have an ident are copied into `local_impl` but are qualified as `type
+///   #ident = <#foreign_path as #disambiguation_path>::#ident;`
+/// * Items that lack an ident are de-duplicated so only unique items that lack an ident are copied
+///   into `local_impl`. Items that lack an ident and also exist verbatim in `local_impl` are not
+///   copied over.
+fn combine_impls(
+	local_impl: ItemImpl,
+	foreign_impl: ItemImpl,
+	foreign_path: Path,
+	disambiguation_path: Path,
+) -> ItemImpl {
+	let existing_local_keys: HashSet<Ident> = local_impl
+		.items
+		.iter()
+		.filter_map(|impl_item| impl_item_ident(impl_item))
+		.cloned()
+		.collect();
+	let existing_unsupported_items: HashSet<ImplItem> = local_impl
+		.items
+		.iter()
+		.filter(|impl_item| impl_item_ident(impl_item).is_none())
+		.cloned()
+		.collect();
+	let mut final_impl = local_impl;
+	let extended_items = foreign_impl.items.into_iter().filter_map(|item| {
+		if let Some(ident) = impl_item_ident(&item) {
+			if existing_local_keys.contains(&ident) {
+				// do not copy colliding items that have an ident
+				None
+			} else {
+				if matches!(item, ImplItem::Type(_)) {
+					// modify and insert uncolliding type items
+					let modified_item: ImplItem = parse_quote! {
+						type #ident = <#foreign_path as #disambiguation_path>::#ident;
+					};
+					Some(modified_item)
+				} else {
+					// copy uncolliding non-type items that have an ident
+					Some(item)
+				}
+			}
+		} else {
+			if existing_unsupported_items.contains(&item) {
+				// do not copy colliding items that lack an ident
+				None
+			} else {
+				// copy uncolliding items without an ident verbaitm
+				Some(item)
+			}
```

Comment:
Same here:

```suggestion
			if existing_unsupported_items.contains(&item) {
				// do not copy colliding items that lack an ident
				return None
			}
			// copy uncolliding items without an ident verbaitm
			Some(item)
```

Reply:
Alternatively, you can even replace this with a couple of new methods:

```suggestion
			// do not copy colliding items that lack an ident
			(!existing_unsupported_items.contains(&item))
				// copy uncolliding items without an ident verbatim
				.then_some(item)
```